OrbitQt/Connections.h

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef ORBIT_QT_CONNECTIONS_H_
+#define ORBIT_QT_CONNECTIONS_H_
+
+#include <memory>
+#include <optional>
+#include <utility>
+
+#include "DeploymentConfigurations.h"
+#include "OrbitBase/Logging.h"
+#include "OrbitClientData/ProcessData.h"
+#include "OrbitClientServices/ProcessManager.h"
+#include "OrbitGgp/Instance.h"
+#include "OrbitSsh/Context.h"
+#include "OrbitSsh/Credentials.h"
+#include "grpcpp/channel.h"
+#include "servicedeploymanager.h"
+
+namespace orbit_qt {
+
+class SshConnectionArtifacts {
+ public:
+  explicit SshConnectionArtifacts(const OrbitSsh::Context* ssh_context,
+                                  ServiceDeployManager::GrpcPort grpc_port,
+                                  const DeploymentConfiguration* deployment_configuration)
+      : ssh_context_(ssh_context),
+        grpc_port_(grpc_port),
+        deployment_configuration_(deployment_configuration) {
+    CHECK(ssh_context != nullptr);
+    CHECK(deployment_configuration != nullptr);
+  }
+
+  [[nodiscard]] const OrbitSsh::Context* GetSshContext() const { return ssh_context_; }
+  [[nodiscard]] const ServiceDeployManager::GrpcPort& GetGrpcPort() const { return grpc_port_; }
+  [[nodiscard]] const DeploymentConfiguration* GetDeploymentConfiguration() const {
+    return deployment_configuration_;
+  }
+
+ private:
+  const OrbitSsh::Context* ssh_context_;
+  const ServiceDeployManager::GrpcPort grpc_port_;
+  const DeploymentConfiguration* deployment_configuration_;
+};
+
+class StadiaConnection {
+  friend class ConnectToStadiaWidget;
```

Comment:
A couple of things come together here. 

On struct vs class, the style guide says the following: "The struct must not have invariants that imply relationships between different fields [...]". When designed as a class, as done here, this is supposed to express to other people , that when this class is instantiated, it is always a valid connection. That's why there are no setter methods to change a field. The fields have relationships between each other in the sense that for example, `service_deploy_manager_` holds the tunnel to a Stadia instance and this is exactly the instance saved in `instance_`. 
StadiaConnection is later moved into MainWindow and used there. MainWindow is not allowed to modify the contents. 

Now when the MainWindow is closed, the StadiaConnection is moved out of MainWindow and the ConnectToStadiaWidget is instantiated again and StadiaConnection is moved into the widget. This is so the user picks up in the ConnectToStadiaWidget, where they left off. A disconnect and reconnect to an instance is not necessary, since StadiaConnection is reused. 

Since ConnectToStadiaWidget needs to be able to only have a partial connection, while establishing the connection, it has its own variables `selected_instance_`, `service_deploy_manager_` and `grpc_channel_`. Now when StadiaConnection is moved into ConnectToStadiaWidget via SetConnection(), the friend access is necessary to move  the members of StadiaConnection into these variables. 

Another consideration was to not use the friend access, but have a `Clear()` method which return the members and then leaves StadiaConnection in a empty state. I decided for the approach via friend, because then StadiaConnection is always valid (when used somewhere else than in ConnectToStadiaWidget).

Does this make sense? 

Ps. I actually just realized that there are no != nullptr checks in the constructor of StadiaConnection. I will add those now

Reply:
Hm, ok, then the class makes sense. I still don't like the usage of `friend` here, but I am also not sure what a resonably better design would be TBH.